pdf/pdfannotation.cpp

```diff
@@ -0,0 +1,356 @@
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; version 2 only.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+#include "pdfannotation.h"
+
+static QString currentAuthor;
+
+class PDFAnnotation::Private
+{
+public:
+    Private(Poppler::Annotation *annotation, PDFDocument *document = nullptr, int page = -1)
+        : owner(document == nullptr || page < 0)
+        , m_annotation(annotation)
+        , m_document(document)
+        , m_page(page)
+    {
+    }
+    ~Private()
+    {
+        if (owner)
+            delete m_annotation;
+    }
+    bool owner;
+    Poppler::Annotation *m_annotation;
+    PDFDocument *m_document;
+    int m_page;
+};
+
+PDFAnnotation::PDFAnnotation(Poppler::Annotation *annotation, QObject *parent)
+    : QObject(parent), d(new PDFAnnotation::Private(annotation))
+{
+    if (annotation && !currentAuthor.isEmpty())
+        annotation->setAuthor(currentAuthor);
+}
+
+PDFAnnotation::PDFAnnotation(Poppler::Annotation *annotation,
+                             PDFDocument *document, int ipage, QObject *parent)
+    : QObject(parent), d(new PDFAnnotation::Private(annotation, document, ipage))
+{
+}
+
+PDFAnnotation::~PDFAnnotation()
+{
+    delete d;
+}
+
+PDFDocument* PDFAnnotation::document() const
+{
+    return d->m_document;
+}
+
+int PDFAnnotation::page() const
+{
+    return d->m_page;
+}
+
+QRectF PDFAnnotation::boundary() const
+{
+    return d->m_annotation->boundary();
+}
+
+void PDFAnnotation::attachOnce(PDFDocument *document, int page)
+{
+    /* Attach annotation only once. */
+    if (d->m_document != nullptr || document == nullptr)
+        return;
+    if (d->m_page >= 0 || page < 0 || page >= document->pageCount())
+        return;
+
+    d->m_document = document;
+    d->m_page = page;
+}
+
+void PDFAnnotation::attach(PDFDocument *document, PDFSelection *selection)
+{
+    if (!selection)
+        return;
+
+    QPair<int, QRectF> word = selection->rectAt(0);
+    attachOnce(document, word.first);
+    
+    QRectF bounds = word.second;
+    for (int i = 1; i < selection->count(); i++)
+        bounds.united(selection->rectAt(i).second);
+
+    d->m_annotation->setBoundary(bounds);
+    d->m_document->addAnnotation(d->m_annotation, d->m_page);
+    emit attached();
+}
+
+void PDFAnnotation::remove()
+{
+    if (d->m_document == nullptr)
+        return;
+
+    d->m_document->removeAnnotation(d->m_annotation, d->m_page);
+    d->m_annotation = nullptr;
+
+    deleteLater();
```

Comment:
Since invokable, could consider checking m_annotation null pointer and always deleting.